### PR TITLE
First prototype for a SpectrumDatasetMakerObs

### PR DESCRIPTION
**Pull request by registerrier**
In the discussion on data reduction, we have proposed to perform a succession of distinct tasks on a per observation basis.

For spectral data reduction the approach is to perform : 
- `SpectrumDataset` creation:
   - creation of the dataset, fill it with counts, aeff, edisp and possibly with the irf background template
- OFF background extraction is needed and creation of a `SpectrumDatasetOnOff`

This PR proposes a prototype for the first step described here. 
It reuses a large fraction of the code of `~gammapy.spectrum.SpectrumExtraction` but removes the loop over observations and the OFF events binning.

To compute the background per energy bin in the spatial region, a spatial integration is necessary. it is performed by creating the background map in a small cutout of the reference tangent map and integrating over pixels inside the region. 

Additionally the capability to perform averaging of the effective area over the region is introduced to better handle the case of extended regions.

### CODE REVIEW

**Comment by adonath** [comment]
Thanks @registerrier! I have thought about a bit what to do about this PR. I agree 100% with the functionality. It's an important use-case, where I know some users are already waiting for it. So it would be nice to get this in even for v0.14.

What I'm not happy yet with is the API and implementation. I think introducing a separate `SpectrumDatasetMakerObs` creates complexity in passing down the configuration e.g. to the `MapMaker`, which is used internally. The other problem I currently see with this implementation is that the counts spectrum is computed from the event list, while the background is computed from a binned map. Another problem I see is that the background norm is currently not adjusted. I think it would be cleaner (concerning configuration and implementation) to introduce either a class or a helper method which creates the `SpectrumDataset` from an existing `MapDataset`. E.g.:

```python3
dataset = MapDataset(...)

# e.g. fit background norm and tilt
fit = Fit(dataset)
fit.run()

maker = SpectrumDatasetMaker(region, spatial_averaging, containment_correction)
spectrum_dataset = maker.run(dataset)
```
 
My hope is that we'll find some time to finish the data reduction PIG and define an API soon, so that we arrive at a flexible maker and pipeline system, where this `SpectrumDatasetMaker` would be one building block e.g. like so: `MapDatasetMaker -> BackgroundFitter -> SpectrumDatasetMaker`. What do you think?

File: gammapy/cube/fit.py [hunk 1]
```diff
@@ -683,6 +688,66 @@ def to_dict(self, filename=""):
         data["filename"] = filename
         return data
 
+    def to_spectrum_dataset(self, on_region):
```
Comment (adonath) [inline gammapy/cube/fit.py]: Maybe also add the `containment_correction` argument here?

Comment (registerrier) [reply gammapy/cube/fit.py]: I have added it here. It requires that the `MapDataset` contains a `PSFMap` as `psf` attribute.

**Comment by registerrier** [comment]
This PR is heavily modified since the initial commit. It now only consists in an addition to `MapDataset` to add a `.to_spectrum_dataset` method to extract a spectrum in any region of a `MapDataset`.
This can then be used to compute average responses over extended sources (so far only average `aeff` and total `background`).

**Comment by adonath** [comment]
Thanks a lot @registerrier! In bdfc0cd I did a little bit of clean-up, I'll go ahead and merge now.

**Pull request merged**